pkg/credentials/filecache.go

```diff
@@ -0,0 +1,230 @@
+	defer func() {
+		_ = lock.Unlock()
+	}()
+	// wait up to a second for the file to lock
+	ctx, cancel := context.WithTimeout(context.TODO(), time.Second)
+	defer cancel()
+	ok, err := lock.TryRLockContext(ctx, 250*time.Millisecond) // try to lock every 1/4 second
+	if !ok {
+		// unable to lock the cache, something is wrong, refuse to use it.
+		return cacheFile{}, fmt.Errorf("unable to read lock file %s: %v", filename, err)
+	}
+	cache := cacheFile{
+		ProfileMap: make(map[string]cachedCredential),
+	}
+	data, err := os.ReadFile(filename)
+	if err != nil {
+		return cache, fmt.Errorf("failed to read cache file: %w", err)
+	}
+	if err := yaml.Unmarshal(data, &cache); err != nil {
+		return cache, fmt.Errorf("unable to parse file %s: %w", filename, err)
+	}
+
+	return cache, nil
+}
+
+// writeCache writes the contents of the credential cache using the
+// yaml marshaled form of the passed cachedCredential object.
+func writeCache(filename string, cache cacheFile) error {
+	lock := flock.New(filename)
+	defer func() {
+		_ = lock.Unlock()
+	}()
+	// wait up to a second for the file to lock
+	ctx, cancel := context.WithTimeout(context.TODO(), time.Second)
+	defer cancel()
+	ok, err := lock.TryRLockContext(ctx, 250*time.Millisecond) // try to lock every 1/4 second
+	if !ok {
+		// unable to lock the cache, something is wrong, refuse to use it.
+		return fmt.Errorf("unable to read lock file %s: %v", filename, err)
+	}
+	data, err := yaml.Marshal(cache)
+	if err == nil {
+		// write privately owned by the user
+		err = os.WriteFile(filename, data, 0600)
+	}
+	return err
+}
+
+// Retrieve implements the Provider interface, returning the cached credential if is not expired,
+// otherwise fetching the credential from the underlying Provider and caching the results on disk
+// with an expiration time.
+func (f *FileCacheProvider) Retrieve() (credentials.Value, error) {
+	if !f.cachedCredential.Expiration.Before(f.clock.Now()) {
+		// use the cached credential
+		return f.cachedCredential.Credential, nil
+	}
+	logger.Info("No cached credential available.  Refreshing...")
+	// fetch the credentials from the underlying Provider
+	credential, err := f.credentials.Get()
+	if err != nil {
+		return credential, err
+	}
+	expiration, err := f.credentials.ExpiresAt()
+	if err != nil {
+		// credential doesn't support expiration time, so can't cache, but still return the credential
+		logger.Warning("Unable to cache credential: %v\n", err)
+		return credential, nil
+	}
+	// underlying provider supports Expirer interface, so we can cache
+	filename, err := cacheFilename()
+	if err != nil {
+		return credential, err
+	}
+	f.cachedCredential = cachedCredential{
+		Credential: credential,
+		Expiration: expiration,
+	}
+	// overwrite whatever was there before. we don't care about multiple creds for various clusters.
+	// if user switches to another role and another profile they have to re-authenticate.
+	cache, _ := readCacheFile(filename)
```

Comment:
err?

Reply:
Doesn't matter at this point. It will just be overwritten anyways. If there was any kind of error, it will create a new file, create a new cache, create a new thing all together.